File: mobile/modules/zetic-mlange/ios/ZeticMlangeModule.swift
import ExpoModulesCore
import Foundation
import ZeticMLange

// Bridges the ZETIC.MLange iOS SDK (1.6.0) into Expo modules. JS sends tensors as
// { shape: [Int], dtype: "int32"|"int64"|"float32", data: base64 }.
//
// The exact Tensor constructor on the SDK may vary between minors. This module uses
// the documented (data: Data, shape: [Int], dataType: DataType) constructor; if your
// SDK exposes a different surface, update buildTensor / extractTensor below.
public class ZeticMlangeModule: Module {
  private var models: [String: ZeticMLangeModel] = [:]
  private let queue = DispatchQueue(label: "lantern.zetic.models", attributes: .concurrent)

  public func definition() -> ModuleDefinition {
    Name("ZeticMlangeModule")

    AsyncFunction("loadModel") { (personalKey: String, modelName: String, version: Int?, modelMode: String) -> String in
      let mode: ModelMode
      switch modelMode.uppercased() {
      case "SPEED": mode = .RUN_SPEED
      case "ACCURACY": mode = .RUN_ACCURACY
      default: mode = .RUN_AUTO
      }
      let model = try ZeticMLangeModel(
        personalKey: personalKey,
        name: modelName,
        version: version as NSNumber?,
        modelMode: mode,
        onDownload: { _ in }
      )
      let handle = UUID().uuidString
      self.queue.async(flags: .barrier) { self.models[handle] = model }
      return handle
    }

    AsyncFunction("runModel") { (handle: String, inputs: [[String: Any]]) -> [[String: Any]] in
      var modelOpt: ZeticMLangeModel?
      self.queue.sync { modelOpt = self.models[handle] }
      guard let model = modelOpt else {
        throw NSError(
          domain: "ZeticMlange",
          code: 1,
          userInfo: [NSLocalizedDescriptionKey: "Unknown model handle: \(handle)"]
        )
      }

      let tensorInputs: [Tensor] = try inputs.map(self.buildTensor)
      let outputs: [Tensor] = try model.run(tensorInputs)
      return outputs.map(self.extractTensor)
    }

    AsyncFunction("releaseModel") { (handle: String) -> Bool in
      var removed = false
      self.queue.async(flags: .barrier) {
        if self.models.removeValue(forKey: handle) != nil { removed = true }
      }
      return removed
    }
  }

  private func buildTensor(_ dict: [String: Any]) throws -> Tensor {
    guard let dtype = (dict["dtype"] as? String)?.lowercased() else {
      throw NSError(domain: "ZeticMlange", code: 2, userInfo: [NSLocalizedDescriptionKey: "Missing dtype"])
    }
    guard let shapeAny = dict["shape"] as? [Any] else {
      throw NSError(domain: "ZeticMlange", code: 3, userInfo: [NSLocalizedDescriptionKey: "Missing shape"])
    }
    let shape: [Int] = shapeAny.compactMap { ($0 as? NSNumber)?.intValue }
    guard let b64 = dict["data"] as? String, let data = Data(base64Encoded: b64) else {
      throw NSError(domain: "ZeticMlange", code: 4, userInfo: [NSLocalizedDescriptionKey: "Invalid data"])
    }
    let dataType: DataType
    switch dtype {
    case "int32": dataType = .INT32
    case "int64": dataType = .INT64
    case "float32": dataType = .FLOAT32
    default:
      throw NSError(domain: "ZeticMlange", code: 5, userInfo: [NSLocalizedDescriptionKey: "Unsupported dtype: \(dtype)"])
    }
    return Tensor(data: data, shape: shape, dataType: dataType)
  }

  private func extractTensor(_ t: Tensor) -> [String: Any] {
    let dtype: String
    switch t.dataType {
    case .INT32: dtype = "int32"
    case .INT64: dtype = "int64"
    case .FLOAT32: dtype = "float32"
    default: dtype = "float32"
    }
    return [
      "shape": t.shape,
      "dtype": dtype,
      "data": t.data.base64EncodedString()
    ]
  }
}

File: mobile/modules/zetic-mlange/android/src/main/java/expo/modules/zeticmlange/ZeticMlangeModule.kt
package expo.modules.zeticmlange

import android.util.Base64
import com.zeticai.mlange.ZeticMLangeModel
import com.zeticai.mlange.ModelMode
import com.zeticai.mlange.core.model.Tensor
import expo.modules.kotlin.exception.CodedException
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import java.nio.ByteBuffer
import java.nio.ByteOrder
import java.util.UUID
import java.util.concurrent.ConcurrentHashMap

// Bridges the ZETIC.MLange Android SDK (1.6.x) into Expo modules. JS sends tensors as
// { shape: int[], dtype: "int32"|"int64"|"float32", data: base64 }.
//
// The exact public Tensor constructor on com.zeticai.mlange.core.model.Tensor may vary
// between SDK minors. This module uses the documented (data: ByteArray, shape: IntArray,
// dtype: DataType) constructor; if your SDK exposes a different constructor (e.g. a
// typed factory), update the `buildTensor` and `extractTensor` helpers below.
class ZeticMlangeModule : Module() {
  private val models = ConcurrentHashMap<String, ZeticMLangeModel>()

  override fun definition() = ModuleDefinition {
    Name("ZeticMlangeModule")

    AsyncFunction("loadModel") { personalKey: String, modelName: String, version: Int?, modelMode: String ->
      val context = appContext.reactContext
        ?: throw CodedException("E_NO_CONTEXT", "Android context unavailable", null)
      val mode = when (modelMode.uppercase()) {
        "SPEED" -> ModelMode.RUN_SPEED
        "ACCURACY" -> ModelMode.RUN_ACCURACY
        else -> ModelMode.RUN_AUTO
      }
      val model = if (version != null) {
        ZeticMLangeModel(context, personalKey, modelName, version, mode) { _ -> }
      } else {
        ZeticMLangeModel(context, personalKey, modelName, modelMode = mode) { _ -> }
      }
      val handle = UUID.randomUUID().toString()
      models[handle] = model
      handle
    }

    AsyncFunction("runModel") { handle: String, inputs: List<Map<String, Any>> ->
      val model = models[handle]
        ?: throw CodedException("E_BAD_HANDLE", "Unknown model handle: $handle", null)
      val tensorInputs: Array<Tensor> = inputs.map(::buildTensor).toTypedArray()
      val outputs: Array<Tensor> = model.run(tensorInputs)
      outputs.map(::extractTensor)
    }

    AsyncFunction("releaseModel") { handle: String ->
      val model = models.remove(handle) ?: return@AsyncFunction false
      try {
        model.deinit()
      } catch (_: Throwable) {}
      true
    }
  }

  private fun buildTensor(map: Map<String, Any>): Tensor {
    val dtype = (map["dtype"] as String).lowercase()
    val shapeAny = map["shape"] as List<*>
    val shape = IntArray(shapeAny.size) { (shapeAny[it] as Number).toInt() }
    val dataB64 = map["data"] as String
    val bytes = Base64.decode(dataB64, Base64.NO_WRAP)
    val buffer = ByteBuffer.wrap(bytes).order(ByteOrder.LITTLE_ENDIAN)
    val dataType = when (dtype) {
      "int32" -> Tensor.DataType.INT32
      "int64" -> Tensor.DataType.INT64
      "float32" -> Tensor.DataType.FLOAT32
      else -> throw CodedException("E_DTYPE", "Unsupported dtype: $dtype", null)
    }
    return Tensor(buffer, shape, dataType)
  }

  private fun extractTensor(t: Tensor): Map<String, Any> {
    val buffer = t.buffer
    buffer.order(ByteOrder.LITTLE_ENDIAN)
    val bytes = ByteArray(buffer.remaining())
    buffer.duplicate().get(bytes)
    val dtype = when (t.dataType) {
      Tensor.DataType.INT32 -> "int32"
      Tensor.DataType.INT64 -> "int64"
      Tensor.DataType.FLOAT32 -> "float32"
      else -> "float32"
    }
    return mapOf(
      "shape" to t.shape.toList(),
      "dtype" to dtype,
      "data" to Base64.encodeToString(bytes, Base64.NO_WRAP)
    )
  }
}
